<center><h1>Técnicas de Inteligencia Artificial</header1></center>

Presentado por: Dario Idrovo Guerrero  <br>
Fecha: 18/07/2022

# Importación de librerias necesarias

In [29]:
#Importación de las siguientes librerías:
import pandas as pd
import numpy as np

#Importar codificador de etiquetas
from sklearn import preprocessing
#Importar libreria de regresion lineal 
from sklearn.linear_model import LinearRegression

## Cargar el Dataset

In [30]:
#Código para cargar el Dataset
detalles = pd.read_csv("CAR_DETAILS.csv")
detalles.head(10)

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type  \
0           ritz  2014           3.35           5.59       27000    Petrol   
1            sx4  2013           4.75           9.54       43000    Diesel   
2           ciaz  2017           7.25           9.85        6900    Petrol   
3        wagon r  2011           2.85           4.15        5200    Petrol   
4          swift  2014           4.60           6.87       42450    Diesel   
5  vitara brezza  2018           9.25           9.83        2071    Diesel   
6           ciaz  2015           6.75           8.12       18796    Petrol   
7        s cross  2015           6.50           8.61       33429    Diesel   
8           ciaz  2016           8.75           8.89       20273    Diesel   
9           ciaz  2015           7.45           8.92       42367    Diesel   

  Seller_Type Transmission  Owner  
0      Dealer       Manual      0  
1      Dealer       Manual      0  
2      Dealer       Manual      0  
3      Dealer       Manual      0  
4      Dealer       Manual      0  
5      Dealer       Manual      0  
6      Dealer       Manual      0  
7      Dealer       Manual      0  
8      Dealer       Manual      0  
9      Dealer       Manual      0

In [31]:
detalles.describe()

Year  Selling_Price  Present_Price     Kms_Driven       Owner
count   301.000000     301.000000     301.000000     301.000000  301.000000
mean   2013.627907       4.661296       7.628472   36947.205980    0.043189
std       2.891554       5.082812       8.644115   38886.883882    0.247915
min    2003.000000       0.100000       0.320000     500.000000    0.000000
25%    2012.000000       0.900000       1.200000   15000.000000    0.000000
50%    2014.000000       3.600000       6.400000   32000.000000    0.000000
75%    2016.000000       6.000000       9.900000   48767.000000    0.000000
max    2018.000000      35.000000      92.600000  500000.000000    3.000000

In [40]:
# objeto label_encoder sabe cómo entender las etiquetas de palabras
label_encoder = preprocessing.LabelEncoder()

# Codificar etiquetas en la columna 'Car_Name', 
# esta nueva etiqueta se denomina 'Car_Types_Name' y se la guarada en una nueva columna
detalles['Car_Name_Trans'] = label_encoder.fit_transform(detalles['Car_Name'])
detalles['Fuel_Type_Trans'] = label_encoder.fit_transform(detalles['Fuel_Type'])
detalles['Seller_Type_Trans'] = label_encoder.fit_transform(detalles['Seller_Type'])
detalles['Transmission_Trans'] = label_encoder.fit_transform(detalles['Transmission'])
detalles.head(10)

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type  \
0           ritz  2014           3.35           5.59       27000    Petrol   
1            sx4  2013           4.75           9.54       43000    Diesel   
2           ciaz  2017           7.25           9.85        6900    Petrol   
3        wagon r  2011           2.85           4.15        5200    Petrol   
4          swift  2014           4.60           6.87       42450    Diesel   
5  vitara brezza  2018           9.25           9.83        2071    Diesel   
6           ciaz  2015           6.75           8.12       18796    Petrol   
7        s cross  2015           6.50           8.61       33429    Diesel   
8           ciaz  2016           8.75           8.89       20273    Diesel   
9           ciaz  2015           7.45           8.92       42367    Diesel   

  Seller_Type Transmission  Owner  Car_Types_Name  Car_Name_Trans  \
0      Dealer       Manual      0              90              90   
1      Dealer       Manual      0              93              93   
2      Dealer       Manual      0              68              68   
3      Dealer       Manual      0              96              96   
4      Dealer       Manual      0              92              92   
5      Dealer       Manual      0              95              95   
6      Dealer       Manual      0              68              68   
7      Dealer       Manual      0              91              91   
8      Dealer       Manual      0              68              68   
9      Dealer       Manual      0              68              68   

   Fuel_Type_Trans  Seller_Type_Trans  Transmission_Trans  
0                2                  0                   1  
1                1                  0                   1  
2                2                  0                   1  
3                2                  0                   1  
4                1                  0                   1  
5                1                  0                   1  
6                2                  0                   1  
7                1                  0                   1  
8                1                  0                   1  
9                1                  0                   1

## Descripción de la fuente del Dataset

Este conjunto de datos contiene información sobre autos usados. Estos datos se pueden usar para muchos propósitos, como la predicción de precios para ejemplificar el uso de la regresión lineal en el aprendizaje automático.

- Variable y
    
    * Selling_Price: Precio de venta del auto


- Variable x
    
    * Car_Name: Nombre del auto
    * Year: Año en que se compro el auto
    * Present_Price: Precio del auto en la sala de exposicion
    * Kms_Driven: Distancia recorrida por el auto en Km
    * Fuel_Type: Tipo de combustible del coche
    * Seller_Type: Tipo de vendedor(Distribuidor/Particular)
    * Transmission: Transmision del auto(Manual/Automatica)
    * Owner: Numero de propietarios que ha tenido del coche

## Explique el problema a resolver. 
Se requiere crear un modelo de aprendizaje automatico, el cual permita predecir el precio adecuado de un automovil en funcion de sus especificaciones. Para realizar se hace uso del dataset descrito en el apartado anterior, el cual contiene 301 samples. Donde, la variable objetivo se denomina **"Selling_Price"**. Existen 8 variables de entradas las cuales son: **Car_Name**,**Year**,**Present_Price**,**Kms_Driven**,**Fuel_Type**,**Seller_Type**,**Transmission** y **Owner**. Cabe destacar que se transforman datos para la implementacion de los procesos que se realicen.

## Caracterización del Dataset

   # Número de instancias en total.

In [41]:
#Se cuentan los registros del dataset.
detalles.count()

Car_Name              301
Year                  301
Selling_Price         301
Present_Price         301
Kms_Driven            301
Fuel_Type             301
Seller_Type           301
Transmission          301
Owner                 301
Car_Types_Name        301
Car_Name_Trans        301
Fuel_Type_Trans       301
Seller_Type_Trans     301
Transmission_Trans    301
dtype: int64

### Número de atributos de entrada, su significado y tipo

   * Car_Name: Nombre del auto
   * Year: Año en que se compro el auto
   * Present_Price: Precio del auto en la sala de exposicion
   * Kms_Driven: Distancia recorrida por el auto en Km
   * Fuel_Type: Tipo de combustible del coche
   * Seller_Type: Tipo de vendedor(Distribuidor/Particular)
   * Transmission: Transmision del auto(Manual/Automatica)
   * Owner: Numero de propietarios que ha tenido del coche

In [42]:
#Muestra la informacion, se incluye el tipo de datos de los 8 atributos de entrada
detalles.drop('Selling_Price', axis=1).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Car_Name            301 non-null    object 
 1   Year                301 non-null    int64  
 2   Present_Price       301 non-null    float64
 3   Kms_Driven          301 non-null    int64  
 4   Fuel_Type           301 non-null    object 
 5   Seller_Type         301 non-null    object 
 6   Transmission        301 non-null    object 
 7   Owner               301 non-null    int64  
 8   Car_Types_Name      301 non-null    int32  
 9   Car_Name_Trans      301 non-null    int32  
 10  Fuel_Type_Trans     301 non-null    int32  
 11  Seller_Type_Trans   301 non-null    int32  
 12  Transmission_Trans  301 non-null    int32  
dtypes: float64(1), int32(5), int64(3), object(4)
memory usage: 24.8+ KB


## Preprocesamiento del dataset. Transformaciones previas necesarias para la modelación

In [43]:
#Variables objetivos y variables de entrada
VariableObjetivo=['Selling_Price']
VariablesEntrada=['Car_Name_Trans', 'Year', 'Present_Price', 'Kms_Driven', 'Fuel_Type_Trans', 'Seller_Type_Trans', 'Transmission_Trans', 'Owner']

#Transformación de las caracteristicas
X=detalles[VariablesEntrada].values
Y=detalles[VariableObjetivo].values

In [44]:
#Estandarización de los datos
from sklearn.preprocessing import StandardScaler
VariablesEntrada_Scaler=StandardScaler()
VariableObjetivo_Scaler=StandardScaler()

In [46]:
#Almacenamiento del objeto de ajuste para referencia posterior
VariablesEntrada_Fit=VariablesEntrada_Scaler.fit(X)
VariableObjetivo_Fit=VariableObjetivo_Scaler.fit(Y)

In [47]:
#Generación de los valores estandarizados de X e y
X=VariablesEntrada_Fit.transform(X)
Y=VariableObjetivo_Fit.transform(Y)

## División del dataset en datos de entrenamiento y datos de test 

In [48]:
#Division de los datos en un conjunto de entrenamiento y prueba
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=42)

In [49]:
#Comprobación rápida del tamaño de los conjuntos de datos de entrenamiento y prueba
print(X_train.shape)
print(Y_train.shape)
print(X_test.shape)
print(Y_test.shape)

(210, 8)
(210, 1)
(91, 8)
(91, 1)


## Regresión Lineal

In [91]:
#Codo de modelo
detalles_r_lineal = LinearRegression()
#Se imprime los parametros de la regrsion lineal
print(detalles_r_lineal)

LinearRegression()


In [92]:
#Se crea el modelo de los datos de entrenamiento
#Uso del modelo, llamando al .fit()
reg_lineal = detalles_r_lineal.fit(X_train,Y_train)
prediction=reg_lineal.predict(X_test)

#Se obtienen resultados con .score()
resul = detalles.score(X_train,Y_train)
#Con el .score usado anteriormente se obtienen los coeficientes de determinación
print(f"Coeficiente de determinación: {resul}")

#Representacion del coeficiente con .intercept_ b0
print(f"Coeficiente Intercepta: {detalles.intercept_}")

#Representacion del coeficiente que representa b1
print(f"Coeficiente Pendiente: {detalles.coef_}")


Coeficiente de determinación: 0.8761476742788223
Coeficiente Intercepta: [0.00614636]
Coeficiente Pendiente: [[-0.00840391  0.20987558  0.73020727 -0.02963663 -0.12194191 -0.1129037
  -0.09669878 -0.05073084]]


In [93]:
#Se crea el modelo de los datos de test
#Uso del modelo, llamando al .fit()
reg_lineal = detalles_r_lineal.fit(X_test,Y_test)
prediction=reg_lineal.predict(X_train)

#Se obtienen resultados con .score()
resul = detalles.score(X_test,Y_test)
#Con el .score usado anteriormente se obtienen los coeficientes de determinación
print(f"Coeficiente de determinación: {resul}")

#Representacion del coeficiente con .intercept_ b0
print(f"Coeficiente Intercepta: {detalles.intercept_}")

#Representacion del coeficiente que representa b1
print(f"Coeficiente Pendiente: {detalles.coef_}")


Coeficiente de determinación: 0.8774173360862039
Coeficiente Intercepta: [0.00614636]
Coeficiente Pendiente: [[-0.00840391  0.20987558  0.73020727 -0.02963663 -0.12194191 -0.1129037
  -0.09669878 -0.05073084]]
